retrieval/discovery/kubernetes/node.go

```diff
@@ -0,0 +1,248 @@
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/util/strutil"
+	"golang.org/x/net/context"
+)
+
+type nodeDiscovery struct {
+	mtx           sync.RWMutex
+	nodes         map[string]*Node
+	retryInterval time.Duration
+	kd            *Discovery
+}
+
+func (d *nodeDiscovery) run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	nodes, _, err := d.getNodes()
+	if err != nil {
+		log.Errorf("Cannot initialize node collection: %s", err)
+	}
+	d.nodes = nodes
+
+	select {
+	case ch <- []*config.TargetGroup{d.updateNodesTargetGroup()}:
+	case <-ctx.Done():
+		return
+	}
+
+	update := make(chan *nodeEvent, 10)
+	go d.watchNodes(update, ctx.Done(), d.retryInterval)
+
+	for {
+		tgs := []*config.TargetGroup{}
+		select {
+		case <-ctx.Done():
+			return
+		case e := <-update:
+			log.Debugf("k8s discovery received node event (EventType=%s, Node Name=%s)", e.EventType, e.Node.ObjectMeta.Name)
+			d.updateNode(e.Node, e.EventType)
+			tgs = append(tgs, d.updateNodesTargetGroup())
+		}
+		if tgs == nil {
+			continue
+		}
+
+		for _, tg := range tgs {
+			select {
+			case ch <- []*config.TargetGroup{tg}:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}
+}
+
+func (d *nodeDiscovery) updateNodesTargetGroup() *config.TargetGroup {
+	d.mtx.RLock()
+	defer d.mtx.RUnlock()
+
+	tg := &config.TargetGroup{
+		Source: nodesTargetGroupName,
+		Labels: model.LabelSet{
+			roleLabel: model.LabelValue("node"),
+		},
+	}
+
+	// Now let's loop through the nodes & add them to the target group with appropriate labels.
+	for nodeName, node := range d.nodes {
+		defaultNodeAddress, nodeAddressMap, err := nodeAddresses(node)
+		if err != nil {
+			log.Debugf("Skipping node %s: %s", node.Name, err)
+			continue
+		}
+
+		kubeletPort := int(node.Status.DaemonEndpoints.KubeletEndpoint.Port)
+
+		address := fmt.Sprintf("%s:%d", defaultNodeAddress.String(), kubeletPort)
+
+		t := model.LabelSet{
+			model.AddressLabel:  model.LabelValue(address),
+			model.InstanceLabel: model.LabelValue(nodeName),
+		}
+
+		for addrType, ip := range nodeAddressMap {
+			labelName := strutil.SanitizeLabelName(nodeAddressPrefix + string(addrType))
+			t[model.LabelName(labelName)] = model.LabelValue(ip[0].String())
+		}
+
+		t[model.LabelName(nodePortLabel)] = model.LabelValue(strconv.Itoa(kubeletPort))
+
+		for k, v := range node.ObjectMeta.Labels {
+			labelName := strutil.SanitizeLabelName(nodeLabelPrefix + k)
+			t[model.LabelName(labelName)] = model.LabelValue(v)
+		}
+		tg.Targets = append(tg.Targets, t)
+	}
+
+	return tg
+}
+
+// watchNodes watches nodes as they come & go.
+func (d *nodeDiscovery) watchNodes(events chan *nodeEvent, done <-chan struct{}, retryInterval time.Duration) {
+	until(func() {
+		nodes, resourceVersion, err := d.getNodes()
+		if err != nil {
+			log.Errorf("Cannot initialize nodes collection: %s", err)
+			return
+		}
+
+		// Reset the known nodes.
+		d.mtx.Lock()
```

Comment:
This is fine, but maybe it'd be simpler to add a new node event type like "clear" or "reset" and emit that down the events channel? Theoretically we can get to lock-free with proper use of channels...